File: Python.PassDetection/src/features/zeroed_position_dominant_foot.py
from typing import List

from src.domain.inferences import Feature
from src.domain.recordings import InputData


class ZeroedPositionDominantFootX(Feature):
    def __init__(self):
        self._values = []

    @property
    def size(self) -> int:
        return 1

    def calculate(self, input_data: InputData) -> None:
        dominant_positions = input_data.user_dominant_foot_positions
        origin = dominant_positions[0]
        self._values = [
            p.x - origin.x for p in dominant_positions
        ]

    @property
    def values(self) -> List[float]:
        return self._values


class ZeroedPositionDominantFootY(Feature):
    def __init__(self):
        self._values = []

    @property
    def size(self) -> int:
        return 1

    def calculate(self, input_data: InputData) -> None:
        dominant_positions = input_data.user_dominant_foot_positions
        origin = dominant_positions[0]
        self._values = [
            p.y - origin.y for p in dominant_positions
        ]

    @property
    def values(self) -> List[float]:
        return self._values


class ZeroedPositionDominantFootZ(Feature):
    def __init__(self):
        self._values = []

    @property
    def size(self) -> int:
        return 1

    def calculate(self, input_data: InputData) -> None:
        dominant_positions = input_data.user_dominant_foot_positions
        origin = dominant_positions[0]
        self._values = [
            p.z - origin.z for p in dominant_positions
        ]

    @property
    def values(self) -> List[float]:
        return self._values

File: Unity.Interactions/Assets/_Project/PassDetection.Replay/Features/ZeroedPositionDominantFoot.cs
using System.Collections.Generic;
using System.Linq;
using _Project.PassDetection.Common;
using Tactive.MachineLearning._Project.MachineLearning;
using Tactive.MachineLearning.Features;
using UnityEngine;

namespace _Project.PassDetection.Replay.Features
{
	public class ZeroedPositionDominantFoot : BaseFeature<InputData>
	{
		public override int Size => 3;

		public override List<Target> Calculate(InputData inputData)
		{
			var dominantPositions = inputData.UserDominantFootPositions;
			var origin = dominantPositions[0];
			var zeroedPositions = dominantPositions
				.Select(p => new Vector3(p.x - origin.x, p.y - origin.y, p.z - origin.z))
				.ToList();

			var xValues = zeroedPositions.Select(pos => pos.x).ToList();
			var yValues = zeroedPositions.Select(pos => pos.y).ToList();
			var zValues = zeroedPositions.Select(pos => pos.z).ToList();

			return new List<Target>
			{
				new Target("zeroed_position_dominant_foot_x", xValues),
				new Target("zeroed_position_dominant_foot_y", yValues),
				new Target("zeroed_position_dominant_foot_z", zValues)
			};
		}
	}
}